imap_processing/lo/l0/science_direct_events.py

```diff
@@ -0,0 +1,202 @@
+            self.tof_decoder = ct.tof_decoder_table[self.case_number]
+
+    def _read_tof_calculation_table(self):
+        """Get the TOF calculation values for this DE's case number."""
+        # The case number determines what calculation value is needed to help
+        # calculate the TOF values
+        tof_calculation_values = ct.tof_calculation_table[self.case_number]
+        self.tof_calculation_binary = {
+            field: bitstring.Bits(calculation_value)
+            for field, calculation_value in tof_calculation_values.items()
+        }
+
+    def _find_decompression_case(self):
+        """Find the decompression case for this DE."""
+        # The first 4 bits of the binary data are used to
+        # determine which case number we are working with.
+        # The case number is used to determine how to
+        # decompress the TOF values.
+        self.case_number = ct.tof_case_table[self.DATA[0:4].bin]
+
+    def _find_remaining_bits(self):
+        """Find which TOF coefficients are needed for a data field and case number."""
+        self.remaining_bits = {}
+        for field, tof_calculation_binary in self.tof_calculation_binary.items():
+            # get a list of tof calculation values as integers
+            # We only need the last 12 bits from the tof calculation binary
+            tof_calculation_array = bitarray(tof_calculation_binary).tolist()[-12:]
+
+            # the tof calculation table binary are used to determine which values
+            # from the tof coefficient table should be used in combination with
+            # the tof decoder.
+            self.remaining_bits[field] = list(
+                compress(ct.tof_coefficient_table, tof_calculation_array)
+            )
+
+    def _parse_binary(self):
+        """Use the TOF decoder to split up the data bits into its TOF fields."""
+        # separate the binary data into its parts
+        self.parsed_bits = {}
+        # the first few bits are only used for determining the case number
+        # so either the first 5 or 6 bits need to be ignored depending the kind
+        # of DE.
+        if self.case_number in [0, 4, 6, 10, 12, 14]:
+            bit_start = 5
+        else:
+            bit_start = 4
+        # Use the TOF decoder to chunk the data binary into its componenets
+        # TOF0, TOF1, TOF2, etc.
+        for field, bit_length in self.tof_decoder.items():
+            self.parsed_bits[field] = self.DATA[bit_start : bit_start + bit_length]
+            bit_start = bit_start + bit_length
+
+    def _decode_fields(self):
+        """Use the parsed data and TOF coefficients to decode the binary."""
+        for field, tof_bits in self.parsed_bits.items():
+            needed_bits = tof_bits[-len(self.remaining_bits[field]) :]
+            # Use the TOF coefficients and the bits for the current field to
+            # calculate the decompressed value. Also round to 2 because the TOF
+            # coefficients only have 2 decimals.
```

Comment:
Are the `needed_bits` and `remaining_bits` floats? Since they are named `bits`, it seems like they are integers at first glance and thus wouldn't need to be rounded until later stages.

Reply:
I definitely need to clean up the naming here. `remaining_bits` are floats which doesn't make any sense. I got that name because a string of bits are used to determine which of those floats should remain, but that name doesn't work well for the resulting list here. I'll update these names to make it more clear